lektor/watcher.py

```diff
@@ -1,60 +1,103 @@
+                path = event.dest_path
             else:
-                self.callback(*item)
+                path = event.src_path
+            self.queue.put((time.time(), event.event_type, path))
+
+
+def _fullname(cls):
+    """Return the full name of a class (including the module name)."""
+    return f"{cls.__module__}.{cls.__qualname__}"
+
+
+def _unique_everseen(seq):
+    """Return the unique elements in sequence, preserving order."""
+    return OrderedDict.fromkeys(seq).keys()
 
 
 class BasicWatcher:
-    def __init__(self, paths, callback=None):
-        self.event_handler = EventHandler(callback=callback)
-        self.observer = Observer()
-        for path in paths:
-            self.observer.schedule(self.event_handler, path, recursive=True)
-        self.observer.setDaemon(True)
+    def __init__(self, paths, observer_classes=(Observer, PollingObserver)):
+        self.event_handler = EventHandler()
+        self.paths = paths
+        self.observer_classes = observer_classes
+        self.observer = None
+
+    def start(self):
+        observer_classes = list(_unique_everseen(self.observer_classes))
+        for observer_class, next_observer_class in zip_longest(
+            observer_classes, observer_classes[1:]
+        ):
+            try:
+                self._start_observer(observer_class)
+                return
+            except Exception as exc:
+                if next_observer_class is None:
+                    raise
+                click.secho(
+                    f"Creation of {_fullname(observer_class)} failed with exception:\n"
+                    f"  {exc.__class__.__name__}: {exc!s}\n"
+                    "This may be due to a configuration or other issue with your system.\n"
+                    f"Falling back to {_fullname(next_observer_class)}.",
+                    fg="red",
+                    bold=True,
+                )
+
+    def stop(self):
+        self.observer.stop()
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, ex_type, ex_value, ex_tb):
+        self.stop()
+
+    def _start_observer(self, observer_class=Observer):
+        if self.observer is not None:
+            raise RuntimeError("Watcher already started.")
+        observer = observer_class()
+        for path in self.paths:
+            observer.schedule(self.event_handler, path, recursive=True)
+        observer.daemon = True
+        observer.start()
+        self.observer = observer
 
     def is_interesting(self, time, event_type, path):
         # pylint: disable=no-self-use
         return True
 
     def __iter__(self):
-        if self.event_handler.queue is None:
-            raise RuntimeError("watcher used with callback")
+        # Alias this since we may need it during interpreter shutdown
+        queue_Empty = queue.Empty
+
         while 1:
             try:
                 item = self.event_handler.queue.get(timeout=1)
```

Comment:
Does anyone have an idea as to why the `timeout` is specified here?

Since timeouts (signaled by `queue.Empty`) are just ignored, it seems to me the timeout could be eliminated (so that the `queue.get` blocks forever).

Reply:
Sounds good.